### DISCUSSION TITLE
Inconsistent / Unexpected $root Behavior?

### DISCUSSION THREAD

**Discussion by Jagadish056** [alpinejs/alpine#4352] [1. Help]
I'm encountering some strange behavior with the `$root` magic property, and I'm hoping to get some clarification on whether this is expected or changeable.

**Example 1**
```
<ul x-data="{ foo() { console.log($root) } }">
    <li x-data x-init="foo">Foo</li>
</ul>
```
_Console Output:_
```
<ul x-data="{ foo() { console.log($root) } }">...</ul>
```
---
**Example 2**
```
<ul x-data="myFoo">
    <li x-data x-init="foo">Foo</li>
</ul>

Alpine.data('myFoo', () => ({
    foo() { console.log(this.$root) }
}))
```
_Console Output:_
```
<li x-data x-init="foo">Foo</li>
```

Even though both examples seem similar in structure, the console output for `$root` is different. In the first example, `$root` refers to the `<ul>` element, while in the second example, `$root` refers to the `<li>` element.

Why, the output of `$root` is different? Is this behavior changeable?

Any insights or explanations would be greatly appreciated!

**Comment by ekwoka** [comment]
Functions on parents are called with the scope of the element they are invoked from, not the element in which they are defined. So the `this.$root` refers to the nearest `x-data` from where the function is called.

This is intended as far as the code itself is written.

I'm not sure if @calebporzio has ever stated explicitly that this is what is expected for all cases.

**Reply by Jagadish056** [reply]
Is there any way to always get the `x-data="myFoo"` element when `foo()` is called? For example, I want to always return the element where `x-data="myFoo"` is assigned (i.e. `<ul>` in the above code). I know there's an `init()` method to assign the `this.$root` to the object variable, but if there's another solution, it might help simplify my code. Any ideas?

**Reply by ekwoka** [reply]
can pass it as an argument to `myFoo` like `myFoo($el)`

The next question becomes "but why?"

**Reply by Jagadish056** [reply]
Not, why!....😁 But, how `<li>` become `<ul>`? I'm expecting something more powerful than the `$root` helper.

Doesn't Alpine automatically detect `$el` from the component instead of a manual pass?

**Reply by ekwoka** [reply]
Yes.

hence in the xdata `myFood($el)` so it would be that element.

> But, how <li> become <ul>? 

I already said.

Function calls are invoked as if from the calling elements scope, not the element they are defined on. the function is in `myFoo` on the `ul`, but the `li` has x-data, and then is it's own `$root`.

> I'm expecting something more powerful than the $root helper.

What are you ACTUALLY trying to do?

**Reply by Jagadish056** [reply]
Hey @ekwoka, you're right. But I'm thinking of handling everything automatically in the background, so there's no need to pass components manually. Since `$el` is available on the component, it should be possible to make this happen.

Just thinking from a dev's perspective. Thoughts?

**Reply by ekwoka** [reply]
I'm not sure what automated means about not being able to pass $el to the data constructor.

A lot of behavior assumes the context is the element calling things.

What is it you are actually trying to do that makes you think you should $root to do it?

**Reply by Jagadish056** [reply]
I'm building an Admin Dashboard with a theme similar to  [SB Admin 2](https://startbootstrap.com/previews/sb-admin-2), using Alpine JS and Tailwind CSS for my upcoming projects. I'm currently working on the sidebar implementation and facing some issues related to how it gonna be displayed on small screens, quick changes on large screens, sidebar scrolling, custom toggle, etc.

Here's a quick demo of what I'm trying to build:
```
<body x-data="toggler">
    <aside x-show="open">
        <ul x-data="sidebar" @scroll="foo" x-resize="foo">
            <li>
                <detail x-data="toggler" @toggle="toggle() && foo()">
                    <summary>Dashboard</summary>
                    <ul x-show="open" x-collapse>
                        <li>Overview</li>
                        <li>Analytics</li>
                        <li>....</li>
                    </ul>
                </detail>
            </li>
        </ul>
    </aside>
</body>
```
https://github.com/user-attachments/assets/baf4beb0-31d6-4dce-ba71-7ca8c528f0d1

**Reply by ekwoka** [reply]
Okay. So what does that have to do with this issue? What does $root have to do with this?

**Reply by Jagadish056** [reply]
`Alpine.magic('selectors', el => (selectors, el=document) => el.querySelectorAll(selectors) )`
```
Alpine.data('sidebar', () => ({
        elem:null,
        init() {
            this.elem = this.$root
        },
        foo() {
            // console.log(this.$root)
            this.$selectors("details", this.elem).forEach(target => {

                const isOpenOnSmallScreen = () => target.open && window.matchMedia('(max-width: 1024px)').matches

                const collapse = this.$selector('ul', target)

                if(!collapse) return

                collapse.style.top = isOpenOnSmallScreen ? `${target.parentNode.offsetTop - this.elem.scrollTop}px` : ''
            });
        }
    }))
```

**Reply by ekwoka** [reply]
Why?

this is just a lot of code to do what a css media query and `display: contents` does for you.

**Reply by Jagadish056** [reply]
I cannot adjust the top position using a CSS media query when the screen size changes in the x and y directions. And, the `<details>` tag feels more comfortable to use for creating collapsible content.

The collapse element is `absolute` to the `body` tag, otherwise inside `aside` it will not work for overflow content on a small screen (< md breakpoint).

Safari, there is no support for `display:contents`. 
![Screenshot 2024-08-26 202442](https://github.com/user-attachments/assets/e68d9349-2080-400b-b29c-8450b1da2721)

**Reply by ekwoka** [reply]
It is supported. Since 11.3 in that chart.

**Reply by Jagadish056** [reply]
Hey @ekwoka,

Thanks for mentioning `display: contents`. I've watched several tutorials on the topic, but I'm still unsure how this CSS style can help simplify my code. Could you provide a demo or example of how to use `display: contents` in this context?

The above code achieves this effect by allowing the sidebar to scroll on small screens (below `1024px`) and quickly reverting to its default position on larger screens (above `1024px`). Additionally, when the screen width is below `480px`, the sidebar automatically hides, with only a custom toggle available to display the sidebar in small screen mode.

Here's a working demo clip showing it in action.

https://github.com/user-attachments/assets/528357c6-104b-46e3-a75a-5c269c4acc51

**Reply by Jagadish056** [reply]
Custom `$root`

```
import defaultAlpine from 'alpinejs'

const Alpine = {
    ...defaultAlpine,
    $root: name => document.querySelector(`[x-data="${name}"]`)
}

let componentName = 'myFoo'

Alpine.data(componentName, () => ({
    foo() {
        console.log(Alpine.$root(componentName))
    }
}))

window.Alpine = Alpine
Alpine.start()
```

**Reply by ekwoka** [reply]
Hard to demo exactly how to use `contents` for your case, but it lets you `break out` of containers for their contents.

And please don't do that.

if you want to add it to `Alpine` just do `Alpine.$root`.

But this code here won't work if there are multiple of that component anyway...

**Reply by Jagadish056** [reply]
Hey @ekwoka,

[findClosest()](https://github.com/alpinejs/alpine/blob/ab743bc5668f018db5d410cad4e48b755696ad9e/packages/alpinejs/src/lifecycle.js#L64C17-L64C28), I've come up with a new sprinkle on the custom `$root`.

Thoughts?
```
(el, name) => Alpine.findClosest(el, el => el.getAttribute('x-data') === name)
```
```
Alpine.$root($el, componentName)
```
